ADR/ADR-003-PyBind11-For-CPP-Bindings.md

```diff
@@ -0,0 +1,295 @@
+ * compile the C++ code into a Python extension module (usually done via the
+   buildsystem, e.g. CMake)
+ * import and use the C++ code directly from Python
+
+PyBind11 offers compatibility with most of C++ `std` (iterators, exceptions,
+smart pointers, variant/tuples, ...) and commonly used modules like NumPy.
+
+
+#### nanobind
+
+[source](https://github.com/wjakob/nanobind) | [documentation](https://nanobind.readthedocs.io/en/latest/)
+
+Licence: BSD-style
+Maturity: new project, start in 2022
+Longevity: 96 contributors (github)
+Activity: very new but high activity (44 Tags)
+Visibility: 3k stars on Github, 485k users
+
+nanobind is similar to PyBind11 (almost the same interface) but with the purpose of
+  
+  * being faster & having less overhead than pybind11
+  * producing smaller binary sizes
+
+while explicitly having reduced compatibility with `std` and third party modules.
+It is not header-only and requires linking to a shared or static library.
+
+
+#### Boost.Python
+
+[source](https://github.com/boostorg/python) | [documentation](https://boostorg.github.io/python/doc/html/reference/index.html)
+
+Licence: Boost Software Licence v1.0
+Maturity: old project, start in 2002
+Longevity: 92 contributors (github)
+Activity: very new but high activity (44 Tags)
+Visibility: 502k stars on Github
+
+Boost.Python is a component of the 
+Boost C++ Libraries designed to simplify interoperability between C++ and Python. 
+
+It provides a high-level, declarative API to expose C++ functions, classes, and objects to Python 
+with minimal boilerplate compared to the raw Python C API.Key aspects:
+
+Integration: Handles type conversions, exception mapping, and reference counting automatically.
+
+C++-centric: Code is written entirely in C++, with bindings expressed through template-based syntax.
+
+Stability: Mature, well-tested, and widely used, though less actively developed than newer alternatives (e.g., pybind11).
+
+Dependencies: Requires the Boost libraries and a compatible C++ compiler.
+
+
+#### SWIG
+
+[source](https://www.swig.org/) | [documentation](https://www.swig.org/doc.html)
+
+Licence: GPLv3, generated code has same licencing as input to SWIG
+Maturity: settled and mature project, initial release 1996
+Longevity: 272 contributors (github)
+Activity: weekly commits, few patches per year
+Visibility: 6.1k stars on Github
+
+SWIG (Simplified Wrapper and Interface Generator) is a tool that automatically generates binding code to expose C/C++ functions and classes 
+to multiple languages, including Python. 
+
+An user-specified interface file (.i) describes what should be exposed, and SWIG generates the C/C++ glue code plus Python wrappers.
+
+Key Points:
+ * Supports many languages besides Python (Java, Ruby, etc.).
+ * Reduces boilerplate by generating the bindings for you.
+ * Handles C++ classes, functions, enums, and more.
+ * Requires SWIG tool installation and a simple build step.
```

Comment:
Make sure you comment on the down sides...

At least where we have used it before, it ends up being horrible...

Reply:
I expanded on the complexity and much lower level abstractions it offers. 

SWIG 3 is leaps and bounds better than 2 but still overkill unless we consider GO/Java bindings. (Because I never ever want to write JNI stuff by hand again...)